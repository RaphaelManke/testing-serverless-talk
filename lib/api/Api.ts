import { LambdaIntegration, RestApi } from "aws-cdk-lib/aws-apigateway";
import { Table } from "aws-cdk-lib/aws-dynamodb";
import { NodejsFunction } from "aws-cdk-lib/aws-lambda-nodejs";
import { Construct } from "constructs";

export class Api extends Construct {
  constructor(scope: Construct, id: string, props: { table: Table }) {
    super(scope, id);

    const api = new RestApi(this, "Api", {});

    const lambda = new NodejsFunction(this, "Lambda", {
      environment: {
        TABLE_NAME: props.table.tableName,
      },
    });
    props.table.grantReadWriteData(lambda);

    const orderResource = api.root.addResource("orders");
    orderResource.addMethod("POST", new LambdaIntegration(lambda));
  }
}
