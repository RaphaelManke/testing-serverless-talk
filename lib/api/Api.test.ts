import { App, Stack } from "aws-cdk-lib";
import { Template } from "aws-cdk-lib/assertions";
import { Api } from "./Api";
import { AttributeType, Table } from "aws-cdk-lib/aws-dynamodb";

it("should have a post order route", () => {
  // Arrange
  const app = new App();
  const testStack = new Stack(app, "testStack");
  const table = new Table(testStack, "Table", {
    partitionKey: { name: "id", type: AttributeType.STRING },
  });
  new Api(testStack, "Api", { table });

  // Act
  const template = Template.fromStack(testStack);

  // Assert
  const apigateway = template.findResources("AWS::ApiGateway::RestApi");
  const orderResource = template.findResources("AWS::ApiGateway::Resource", {
    Properties: {
      PathPart: "orders",
      RestApiId: {
        Ref: Object.keys(apigateway)[0],
      },
    },
  });
  expect(orderResource).toMatchSnapshot();
  template.findResources("AWS::ApiGateway::Method", {
    Properties: {
      HttpMethod: "POST",
      RestApiId: {
        Ref: Object.keys(apigateway)[0],
      },
      ResourceId: {
        Ref: Object.keys(orderResource)[0],
      },
    },
  });
});

it("should give lambda access to table", () => {
  // Arrange
  const app = new App();
  const testStack = new Stack(app, "testStack");
  const table = new Table(testStack, "Table", {
    partitionKey: { name: "id", type: AttributeType.STRING },
  });
  const grantWriteSpy = jest.spyOn(table, "grantReadWriteData");
  new Api(testStack, "Api", { table });

  // Act
  const template = Template.fromStack(testStack);

  // Assert
  expect(grantWriteSpy).toBeCalledTimes(1);

  const lambda = template.findResources("AWS::Lambda::Function");
  expect(Object.values(lambda)[0].Properties.Environment)
    .toMatchInlineSnapshot(`
{
  "Variables": {
    "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
    "TABLE_NAME": {
      "Ref": "TableCD117FA1",
    },
  },
}
`);
});
