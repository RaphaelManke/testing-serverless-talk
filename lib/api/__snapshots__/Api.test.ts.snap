// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should have a post order route 1`] = `
{
  "Apiorders48D17054": {
    "Properties": {
      "ParentId": {
        "Fn::GetAtt": [
          "ApiCD79AAA0",
          "RootResourceId",
        ],
      },
      "PathPart": "orders",
      "RestApiId": {
        "Ref": "ApiCD79AAA0",
      },
    },
    "Type": "AWS::ApiGateway::Resource",
  },
}
`;
